xwiki-platform-distribution/xwiki-platform-distribution-migrations/src/main/java/org/xwiki/migrations/internal/R170400000XWIKI23160DataMigration.java

```diff
@@ -0,0 +1,175 @@
+import org.xwiki.component.annotation.Component;
+import org.xwiki.extension.ExtensionId;
+import org.xwiki.extension.repository.InstalledExtensionRepository;
+import org.xwiki.extension.xar.internal.handler.XarExtensionHandler;
+import org.xwiki.extension.xar.internal.handler.packager.Packager;
+import org.xwiki.extension.xar.internal.repository.XarInstalledExtension;
+import org.xwiki.extension.xar.internal.repository.XarInstalledExtensionRepository;
+import org.xwiki.model.reference.DocumentReference;
+import org.xwiki.model.reference.LocalDocumentReference;
+import org.xwiki.xar.XarEntry;
+import org.xwiki.xar.XarException;
+
+import com.xpn.xwiki.XWiki;
+import com.xpn.xwiki.XWikiContext;
+import com.xpn.xwiki.XWikiException;
+import com.xpn.xwiki.doc.XWikiDocument;
+import com.xpn.xwiki.internal.mandatory.XWikiPreferencesDocumentInitializer;
+import com.xpn.xwiki.objects.BaseObject;
+import com.xpn.xwiki.store.migration.DataMigrationException;
+import com.xpn.xwiki.store.migration.XWikiDBVersion;
+import com.xpn.xwiki.store.migration.hibernate.AbstractHibernateDataMigration;
+
+import static com.xpn.xwiki.internal.mandatory.XWikiPreferencesDocumentInitializer.LOCAL_REFERENCE;
+
+/**
+ * Empty fields from {@link XWikiPreferencesDocumentInitializer#LOCAL_REFERENCE} used as templates when they are
+ * replaced by actual velocity templates.
+ *
+ * @version $Id$
+ * @since 17.4.0RC1
+ */
+@Component
+@Named("R170400000XWIKI23160")
+@Singleton
+public class R170400000XWIKI23160DataMigration extends AbstractHibernateDataMigration
+{
+    private static final String META_FIELD = "meta";
+
+    @Inject
+    private Packager packager;
+
+    @Inject
+    @Named(XarExtensionHandler.TYPE)
+    private InstalledExtensionRepository installedExtensionRepository;
+
+    @Override
+    public XWikiDBVersion getVersion()
+    {
+        return new XWikiDBVersion(170400000);
+    }
+
+    @Override
+    public String getDescription()
+    {
+        return "Cleanup XWikiProperties fields used as templates when they match the default values.";
+    }
+
+    @Override
+    protected void hibernateMigrate() throws DataMigrationException, XWikiException
+    {
+        // TODO: confirmation_email_content and validation_email_content fields also need to be migrated, see
+        //  XWIKI-23164
+        XWikiContext xWikiContext = getXWikiContext();
+        XWiki wiki = xWikiContext.getWiki();
+        // Defining the root local is important to have entity equality when retrieving the version from the XAR.
+        DocumentReference xwikiPreferencesDocumentReference =
+            new DocumentReference(new LocalDocumentReference(LOCAL_REFERENCE, Locale.ROOT),
+                xWikiContext.getWikiReference());
+        // We manipulate a cloned version to avoid conflicts with potential concurrent modifications.
+        XWikiDocument xwikiPreferencesDocument =
+            wiki.getDocument(xwikiPreferencesDocumentReference, xWikiContext).clone();
```

Comment:
Minor: you might want to clone `xwikiPreferencesDocument` only if you need to modify it (even if, I guess, it will be the case most of the time).

Reply:
I believe it's a marginal cost. I'll think of a solution that does not make the code less readable though.